### ISSUE TITLE
AKHQ connection with KEYCLOAK - Configure problem

### ISSUE THREAD

**Issue by ericmagna** [tchiotludo/akhq#607]
I am trying to configure AKHQ for a login with the keycloak provider. But I am encountering an error which I cannot understand this behavior.

![page1](https://user-images.githubusercontent.com/69892938/108171322-a6a70d00-70fb-11eb-8663-eac3e5c5a5cc.png)
![page2](https://user-images.githubusercontent.com/69892938/108171335-aa3a9400-70fb-11eb-9843-4bf31e6f1c93.png)
When I am on the home page, I click on the login button which takes me to the page with a new "Login with oidc" button. I click on this button and I find an errored page on a redirection problem (http status code 307) with a loop error.

I have tried multiple configurations, the last of which you will find attached. You may have encountered this anomaly. I have probably misconfigured the application.yml file to connect it to the keycloak but I cannot find the syntax in error causing the page redirection. Do you have an idea?

thankin
![page3](https://user-images.githubusercontent.com/69892938/108171378-ba527380-70fb-11eb-8ec7-2c0917f34496.png)
![page4](https://user-images.githubusercontent.com/69892938/108171393-be7e9100-70fb-11eb-8bb0-c9ebcc86259f.png)
[application.txt](https://github.com/tchiotludo/akhq/files/5993994/application.txt)

g you for your help.

**Comment by ericmagna** [comment]
I am trying to configure AKHQ for a login with the keycloak provider. But I am encountering an error which I cannot understand this behavior.

When I am on the home page, I click on the login button which takes me to the page with a new "Login with oidc" button. I click on this button and I find an errored page on a redirection problem (http status code 307) with a loop error.

I have tried multiple configurations, the last of which you will find attached. You may have encountered this anomaly. I have probably misconfigured the application.yml file to connect it to the keycloak but I cannot find the syntax in error causing the page redirection. Do you have an idea?

i change application.yml to application.txt to upload file.

thanking you for your help.

**Comment by ericmagna** [comment]
i change application.yml to application.txt to upload file.

thanking you for your help to try a solution.

**Comment by alwibrm** [comment]
Have you tried without the properties `authorization` and `token`? So only:
```
    # OIDC authentification configuration
    oauth2:
      enabled: true
      clients:
        grant-type: password
        client-id: kafka-client
        client-secret: <secret>
        openid:
          issuer: https://<url keycloak>/auth/realms/sandbox_kafka
```

**Comment by ericmagna** [comment]
Hello,

yes I made this configuration.

To be sure, I started this configuration again.
I have attached 2 retested configurations.
I always have the same behavior.
I can't trace a connection between AKHQ and Keycloak but I can't understand this abnormal behavior.

Thank you

[application.yml.save8.txt](https://github.com/tchiotludo/akhq/files/5997661/application.yml.save8.txt)
[application.yml.save9.txt](https://github.com/tchiotludo/akhq/files/5997662/application.yml.save9.txt)

**Comment by tchiotludo** [comment]
* Maybe you can try to increase log level : 

```bash
curl -i -X POST -H "Content-Type: application/json" \
       -d '{ "configuredLevel": "TRACE" }' \
       http://localhost:8081/loggers/io.micronaut.security
       
       
curl -i -X POST -H "Content-Type: application/json" \
       -d '{ "configuredLevel": "TRACE" }' \
       http://localhost:8081/loggers/org.akhq.configs
```
* In my opinion, remove all the default config from your application.yml, it will help to have feedback, I notice this part that seems to be strange : 
```yaml
        signatures:
          jwks:
            keycloak:
              url: https://<url keycloak>/auth/realms/sandbox_kafka/protocol/openid-connect/certs
```
* what version of akhq are you using ?

**Comment by ericmagna** [comment]
I redid the configuration by taking your comments and I manage to see calls with the keycloak but in error (see page5.png). This anomaly is a configuration problem of the application.yml file.

In your opinion, in this error, is a section missing in the configuration?

When I click on the login button, I arrive on a page with a "login oidc" button. I thought I would find a text box to enter the user and another text box to enter the password. In the sample setup, I can't figure out how to do this. Could you bring me how I should do?

Thank for you help.
[trace.zip](https://github.com/tchiotludo/akhq/files/5998826/trace.zip)

**Comment by hoeggi** [comment]
The error message in `page5.png` in your `trace.zip` means the redirect urls in keycloak are not configured correctly. Look at the address bar and check what the query parameter for `redirect_uri` is and configure that url in your keycloak client.

**Comment by hoeggi** [comment]
About the redirect loop, I also see this sometimes when I' redirected to `/login` instead of `/ui/login` but didn't have time yet to check why that is (also this loop happens with basic auth too).

**Comment by tchiotludo** [comment]
I've try look at logs, I found this lines : 
```
2021-02-17 21:15:12,823 [39mDEBUG[0;39m [35mmain      [0;39m [36mEndSessionEndpointResolver[0;39m Resolving the end session endpoint for provider [oidc]. Looking for a bean with the provider name qualifier
2021-02-17 21:15:12,834 [39mDEBUG[0;39m [35mmain      [0;39m [36mEndSessionEndpointResolver[0;39m No EndSessionEndpoint bean found with a name qualifier of [oidc]
2021-02-17 21:15:12,835 [39mDEBUG[0;39m [35mmain      [0;39m [36mEndSessionEndpointResolver[0;39m No EndSessionEndpoint can be resolved. The issuer for provider [oidc] does not match any of the providers supported by default
2021-02-17 21:15:12,946 [39mDEBUG[0;39m [35mmain      [0;39m [36m.m.s.o.r.OauthRouteBuilder[0;39m Registering login route [GET: /oauth/login/oidc] for oauth configuration [oidc]
2021-02-17 21:15:12,951 [39mDEBUG[0;39m [35mmain      [0;39m [36m.m.s.o.r.OauthRouteBuilder[0;39m Registering callback route [GET: /oauth/callback/oidc] for oauth configuration [oidc]
2021-02-17 21:15:12,953 [39mDEBUG[0;39m [35mmain      [0;39m [36m.m.s.o.r.OauthRouteBuilder[0;39m Registering callback route [POST: /oauth/callback/oidc] for oauth configuration [oidc]
2021-02-17 21:15:12,954 [39mDEBUG[0;39m [35mmain      [0;39m [36m.m.s.o.r.OauthRouteBuilder[0;39m Skipped registration of logout route. No openid clients found that support end session
```

Specially this `No EndSessionEndpoint can be resolved.`, leading to [this](https://github.com/micronaut-projects/micronaut-security/blob/master/security-oauth2/src/main/java/io/micronaut/security/oauth2/endpoint/endsession/request/EndSessionEndpointResolver.java#L88) where no support for keycloak. 
I think it can work looking at this [tutorial](https://piotrminkowski.com/2020/09/21/micronaut-oauth2-and-security-with-keycloak/) but there is a lot of configuration to be done. 

Can you try please ?

**Comment by ericmagna** [comment]
Thank you for your comments. I look at the configuration and the link to the article.
I am new to keycloak, I need to understand the configuration as well as possible.

**Comment by ericmagna** [comment]
hello,

i got to connect akhq with keycloak. I arrive on the home page (page1.png) and I click on login to arrive at the button to authenticate to keycloak (page2.png). Then I arrive on the keycloak to authenticate with the user / password (page3.png). I validate and I am directed to the login page (page4.png).

I always stay on the same page even if I type the url http://10.207.69.185:18080 which directs me to the login page (page4.png).

In the keycloak trace file (keycloak.txt trace) I read the effective authentication and in the stdout.log file I see the authentication well.

In the akhq configuration (application.yml), I put the definition on my user in admin. If I understand correctly when I authenticate with my user, I should be logged in in admin mode. However, I always stay on the same page (page4.png).
I cannot understand the behavior of the akhq controller. I put in the attached file the configuration in the keycloak in case you have been confronted with this configuration with the keycloak to get the information.

I voluntarily remove the notion of group to stay on an authentication rule on a specific user.

I'm running out of ideas.

If you have a lead, I'm interested.

Thank you for your help.

[trace.zip](https://github.com/tchiotludo/akhq/files/6018291/trace.zip)

**Comment by tchiotludo** [comment]
Not really sure since I don't do the implementation by myself for the openIdConnect part. 
Seems that there is some misunderstanding on this part. 

Mostly this issue https://github.com/tchiotludo/akhq/issues/534 talk a lot about it, and this PR want to clarify it : https://github.com/tchiotludo/akhq/pull/573.

As I understand, you need to : 
* add this : ` groups-field: roles`, that will return the field where you have some groups on KeyCloak 
* add some mapping between your groups with akhq group 
 ```yaml
            - name: mathematicians
              groups:	              groups:
                # the corresponding akhq groups (eg. topic-reader/writer or akhq default groups like admin/reader/no-role)
                - topic-reader	                - topic-reader
```

I see that you have some roles on your log : 
`roles=>["adminadmin","adminuma_authorization","adminoffline_access"]`

Maybe you can try this : 
```yaml
    oidc:
      enabled: true
      providers:
        oidc:
          label: "Login with Keycloak"
          username-field: preferred_username
          default-group: reader
          groups-field: roles
          groups:
            - name: adminadmin	            
              groups:
                - admin
```

Looking at the token here : https://jwt.io/
Seems the token is clean, and you are logged in micronaut, but mapping between akhq and micronaut don't work. 

Mostly I think looking at the code, the default field for group are `roles` and since you have the field on your token, maybe we map the group, no matter the user, ... not really sure about that ... 

Also there is an issue with 0.16 auth part that could affect you ? , are you using the last version (0.17.0) ? 

Sorry doing blindness, I don't have the time to setup a full keycloak in order to test by myself

**Comment by ericmagna** [comment]
Hello,

I am using version 0.17.

Thank you for your comments, they will help move the subject forward with keycloak.

thank you.

**Comment by ericmagna** [comment]
> About the redirect loop, I also see this sometimes when I'm redirected to `/login` instead of `/ui/login` but didn't have time yet to check why that is (also this loop happens with basic auth too).



> I've try look at logs, I found this lines :
> 
> ```
> 2021-02-17 21:15:12,823 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36mEndSessionEndpointResolver�[0;39m Resolving the end session endpoint for provider [oidc]. Looking for a bean with the provider name qualifier
> 2021-02-17 21:15:12,834 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36mEndSessionEndpointResolver�[0;39m No EndSessionEndpoint bean found with a name qualifier of [oidc]
> 2021-02-17 21:15:12,835 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36mEndSessionEndpointResolver�[0;39m No EndSessionEndpoint can be resolved. The issuer for provider [oidc] does not match any of the providers supported by default
> 2021-02-17 21:15:12,946 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36m.m.s.o.r.OauthRouteBuilder�[0;39m Registering login route [GET: /oauth/login/oidc] for oauth configuration [oidc]
> 2021-02-17 21:15:12,951 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36m.m.s.o.r.OauthRouteBuilder�[0;39m Registering callback route [GET: /oauth/callback/oidc] for oauth configuration [oidc]
> 2021-02-17 21:15:12,953 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36m.m.s.o.r.OauthRouteBuilder�[0;39m Registering callback route [POST: /oauth/callback/oidc] for oauth configuration [oidc]
> 2021-02-17 21:15:12,954 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36m.m.s.o.r.OauthRouteBuilder�[0;39m Skipped registration of logout route. No openid clients found that support end session
> ```
> 
> Specially this `No EndSessionEndpoint can be resolved.`, leading to [this](https://github.com/micronaut-projects/micronaut-security/blob/master/security-oauth2/src/main/java/io/micronaut/security/oauth2/endpoint/endsession/request/EndSessionEndpointResolver.java#L88) where no support for keycloak.
> I think it can work looking at this [tutorial](https://piotrminkowski.com/2020/09/21/micronaut-oauth2-and-security-with-keycloak/) but there is a lot of configuration to be done.
> 
> Can you try please ?



I read issue #574 and i

> I've try look at logs, I found this lines :
> 
> ```
> 2021-02-17 21:15:12,823 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36mEndSessionEndpointResolver�[0;39m Resolving the end session endpoint for provider [oidc]. Looking for a bean with the provider name qualifier
> 2021-02-17 21:15:12,834 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36mEndSessionEndpointResolver�[0;39m No EndSessionEndpoint bean found with a name qualifier of [oidc]
> 2021-02-17 21:15:12,835 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36mEndSessionEndpointResolver�[0;39m No EndSessionEndpoint can be resolved. The issuer for provider [oidc] does not match any of the providers supported by default
> 2021-02-17 21:15:12,946 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36m.m.s.o.r.OauthRouteBuilder�[0;39m Registering login route [GET: /oauth/login/oidc] for oauth configuration [oidc]
> 2021-02-17 21:15:12,951 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36m.m.s.o.r.OauthRouteBuilder�[0;39m Registering callback route [GET: /oauth/callback/oidc] for oauth configuration [oidc]
> 2021-02-17 21:15:12,953 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36m.m.s.o.r.OauthRouteBuilder�[0;39m Registering callback route [POST: /oauth/callback/oidc] for oauth configuration [oidc]
> 2021-02-17 21:15:12,954 �[39mDEBUG�[0;39m �[35mmain      �[0;39m �[36m.m.s.o.r.OauthRouteBuilder�[0;39m Skipped registration of logout route. No openid clients found that support end session
> ```
> 
> Specially this `No EndSessionEndpoint can be resolved.`, leading to [this](https://github.com/micronaut-projects/micronaut-security/blob/master/security-oauth2/src/main/java/io/micronaut/security/oauth2/endpoint/endsession/request/EndSessionEndpointResolver.java#L88) where no support for keycloak.
> I think it can work looking at this [tutorial](https://piotrminkowski.com/2020/09/21/micronaut-oauth2-and-security-with-keycloak/) but there is a lot of configuration to be done.
> 
> Can you try please ?

**Comment by ericmagna** [comment]
wrong action

**Comment by ericmagna** [comment]
I added the group configuration and I always come back to the authentication page.

I have read issue #573 (#534) with a code modification explanation. My akhq setup seems to match. On the other hand, I wonder about the configuration of the keycloak providers. Indeed, reading hoeggi's messages shows the use of keycloak but I can't find enough information on how he configured the keycloak provider part. It can be configured the provider so that the AKHQ configuration can correctly interpret it. Do you have any information on this point ?

Thanking you.

Config security :
 # OIDC configuration
    oidc:
      enabled: true
      providers:
        oidc:
          label: "Login with Keycloak"
          username-field: preferred_username
          groups-field: roles
          default-group: reader
          groups:
            - name: adminadmin
              groups:
                - admin
          users:
            - username: ac75006583
              groups:
                - admin

**Comment by ericmagna** [comment]
change value for authentification to cookie and delete this section 
    token:
      jwt:
        enabled: true
        signatures.jwks.keycloak:
          url: <url keycloak>

I configured the provider for authentication only on a user.
After restarting the service, it works.

thank for you help.

**Comment by tchiotludo** [comment]
Glad to know that, maybe you can create a pull request with a full example on the readme please ?
It will help people I think !

**Comment by ericmagna** [comment]
I noted to add Keycloak configuration feedback.

I haven't finished yet because I still have 2 points.

The 1st point is once logged in, when I click on logout, I always come back to the login page. The logout function is not working properly because I must have the wrong configuration.

The second point is not to have done a test with the notion of group.

I looked for an issue corresponding to a problem on logout but I found nothing.
I continue to search by checking parameters.

thank.
